espresso/streamer_test.go

```diff
@@ -0,0 +1,589 @@
+// environment.
+//
+// As we progress through the tests, we should be able to update our local mock
+// state, and then perform our various `.Update` and `.Next` calls, in order to
+// verify that we end up with the expected state.
+//
+// The current expected use case for the Streamer is for the user to "Refresh"
+// the state of the streamer by calling `.Refresh`.
+type MockStreamerSource struct {
+	// At the moment the Streamer utilizes the SyncStatus in order to update
+	// it's local state.  But, in general the Streamer doesn't consume all
+	// of the fields provided within the SyncStatus.  At the moment it only
+	// cares about SafeL2, and FinalizedL1. So this is what we will track
+
+	FinalizedL1 eth.L1BlockRef
+	SafeL2      eth.L2BlockRef
+
+	EspTransactionData map[EspBlockAndNamespace]esp_client.TransactionsInBlock
+	LatestEspHeight    uint64
+}
+
+// AdvanceFinalizedL1ByNBlocks advances the FinalizedL1 block reference by n blocks.
+func (m *MockStreamerSource) AdvanceFinalizedL1ByNBlocks(n uint) {
+	m.FinalizedL1 = createL1BlockRef(m.FinalizedL1.Number + uint64(n))
+}
+
+// AdvanceFinalizedL1 advances the FinalizedL1 block reference by one block.
+func (m *MockStreamerSource) AdvanceFinalizedL1() {
+	m.FinalizedL1 = createL1BlockRef(m.FinalizedL1.Number + 1)
+}
+
+// AdvanceL2ByNBlocks advances the SafeL2 block reference by n blocks.
+func (m *MockStreamerSource) AdvanceL2ByNBlocks(n uint) {
+	m.SafeL2 = createL2BlockRef(m.SafeL2.Number+uint64(n), m.FinalizedL1)
+}
+
+// AdvanceSafeL2 advances the SafeL2 block reference by one block.
+func (m *MockStreamerSource) AdvanceSafeL2() {
+	m.SafeL2 = createL2BlockRef(m.SafeL2.Number+1, m.FinalizedL1)
+}
+
+// AdvanceEspressoHeightByNBlocks advances the LatestEspHeight by n blocks.
+func (m *MockStreamerSource) AdvanceEspressoHeightByNBlocks(n int) {
+	m.LatestEspHeight += uint64(n)
+}
+
+// AdvanceEspressoHeight advances the LatestEspHeight by one block.
+func (m *MockStreamerSource) AdvanceEspressoHeight() {
+	m.LatestEspHeight++
+}
+
+// SyncStatus returns the current sync status of the mock streamer source.
+// Only the fields FinalizedL1, FinalizedL1, and SafeL2 are populated, as those
+// are the only fields explicitly inspected by the EspressoStreamer.
+func (m *MockStreamerSource) SyncStatus() *eth.SyncStatus {
+	return &eth.SyncStatus{
+		FinalizedL1: m.FinalizedL1,
+		SafeL2:      m.SafeL2,
+	}
+}
+
+func (m *MockStreamerSource) AddEspressoTransactionData(height, namespace uint64, txData esp_client.TransactionsInBlock) {
+	if m.EspTransactionData == nil {
+		m.EspTransactionData = make(map[EspBlockAndNamespace]esp_client.TransactionsInBlock)
+	}
+
+	m.EspTransactionData[BlockAndNamespace(height, namespace)] = txData
+
+	if m.LatestEspHeight < height {
+		m.LatestEspHeight = height
+	}
+}
+
+var _ espresso.L1Client = (*MockStreamerSource)(nil)
+
+// L1 Client methods
+
+func (m *MockStreamerSource) HeaderHashByNumber(ctx context.Context, number *big.Int) (common.Hash, error) {
```

Comment:
Maybe rename the function to `L1HeaderHashByNumber`?

Reply:
In this case, sadly not.  This method exists to adhere to the interface that the `EspressoStreamer` is consuming from.  So changing the name would mean it no longer adheres to the interface.

We could change the name, but it's something @dailinsubjam named, and it would require us to modify it in the three places that it is referenced.  Though, this may be worth while since it's all code we introduced, and it helps to clarify the specific functionality.

Reply:
Sure, let us keep it as is.